File: templates/problems/07_trees/08_level_order/python/solution.py
from helpers import TreeNode, Optional


def solve_level_order(root: Optional[TreeNode]) -> list[list[int]]:
    """Return the level order traversal of a binary tree's values.

    Time: O(n), Space: O(n).
    """
    if root is None:
        return []
    result: list[list[int]] = []
    queue: list[TreeNode] = [root]
    while queue:
        size = len(queue)
        level: list[int] = []
        for _ in range(size):
            node = queue.pop(0)
            level.append(node.val)
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)
        result.append(level)
    return result

File: templates/problems/07_trees/08_level_order/go/solution.go
package level_order

// solveLevelOrder returns the level order traversal of a binary tree's values.
// Time: O(n), Space: O(n).
func solveLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var result [][]int
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		size := len(queue)
		level := make([]int, 0, size)
		for i := 0; i < size; i++ {
			node := queue[i]
			level = append(level, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		queue = queue[size:]
		result = append(result, level)
	}
	return result
}
